File: src/content/docs/topics/cs/coding-problems/tries/208-implement-trie-approach2.go
package main

import "fmt"

type TrieNode struct {
	Children [26]*TrieNode
	IsEnd    bool
}

type Trie struct {
	Root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{Root: &TrieNode{}}
}

func (t *Trie) Insert(word string) {
	node := t.Root
	for _, ch := range word {
		i := ch - 'a'
		if node.Children[i] == nil {
			node.Children[i] = &TrieNode{}
		}
		node = node.Children[i]
	}
	node.IsEnd = true
}

func (t *Trie) Search(word string) bool {
	node := t.walk(word)
	return node != nil && node.IsEnd
}

func (t *Trie) StartsWith(prefix string) bool {
	return t.walk(prefix) != nil
}

func (t *Trie) walk(s string) *TrieNode {
	node := t.Root
	for _, ch := range s {
		i := ch - 'a'
		if node.Children[i] == nil {
			return nil
		}
		node = node.Children[i]
	}
	return node
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func runTests() {
	t := NewTrie()
	t.Insert("apple")
	assert(t.Search("apple") == true)
	assert(t.Search("app") == false)
	assert(t.StartsWith("app") == true)
	t.Insert("app")
	assert(t.Search("app") == true)
	assert(t.Search("ap") == false)
	assert(t.StartsWith("b") == false)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/tries/208-implement-trie-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class TrieNode {
    children: Map<string, TrieNode> = new Map();
    isEnd: boolean = false;
}

class Trie {
    private root = new TrieNode();

    insert(word: string): void {
        let node = this.root;
        for (const ch of word) {
            if (!node.children.has(ch))
                node.children.set(ch, new TrieNode());
            node = node.children.get(ch)!;
        }
        node.isEnd = true;
    }

    search(word: string): boolean {
        const node = this._walk(word);
        return node !== null && node.isEnd;
    }

    startsWith(prefix: string): boolean {
        return this._walk(prefix) !== null;
    }

    private _walk(s: string): TrieNode | null {
        let node = this.root;
        for (const ch of s) {
            if (!node.children.has(ch)) return null;
            node = node.children.get(ch)!;
        }
        return node;
    }
}

const trie = new Trie();
trie.insert("apple");
assert(trie.search("apple") === true);
assert(trie.search("app") === false);
assert(trie.startsWith("app") === true);
trie.insert("app");
assert(trie.search("app") === true);
assert(trie.search("ap") === false);
assert(trie.startsWith("b") === false);
console.log("all tests pass");
